### ISSUE TITLE
How to generate heatmap for yolo11 segmentation task?

### ISSUE THREAD

**Issue by AnikiFan** [ultralytics/ultralytics#17142]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi! How can I get probability heatmap for yolo11 segmentation task for further process?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @AnikiFan, thank you for reaching out and your interest in Ultralytics 🚀! We recommend starting by exploring the [Docs](https://docs.ultralytics.com) for guidance and insights where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. These resources might already contain the answers you're looking for regarding generating heatmaps for YOLO11 segmentation tasks.

If your question is related to custom training or a specific feature request, please provide as much detail as possible, including any relevant data or code snippets you are using. 🤓 This is especially helpful if you're encountering a unique challenge.

Since this is an automated response, an engineer from Ultralytics will review your issue and provide further assistance soon. In the meantime, feel free to engage with the Ultralytics community on platforms like [Discord](https://ultralytics.com/discord) 🎧, [Discourse](https://community.ultralytics.com), or [Subreddit](https://reddit.com/r/ultralytics) to exchange ideas and solutions.

## Upgrade

Ensure your `ultralytics` package is up to date, which might help you access the latest features and resolve existing issues:

```bash
pip install -U ultralytics
```

## Environments

If you want to run your tasks in a verified environment with preinstalled dependencies, you can try any of the following:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. Refer to the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. Check out the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. Follow the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

You can monitor the ongoing status of our CI tests to ensure all setups, including YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/), pass on various platforms such as macOS, Windows, and Ubuntu every 24 hours.

We appreciate your patience and look forward to helping you further!

**Comment by Y-T-G** [comment]
Do you mean mask?

**Comment by AnikiFan** [comment]
kind of，I mean the the matrix used to get 01mask by tresholding

**Comment by Y-T-G** [comment]
The masks can be accessed by indexing `result.masks.data`

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #25548](https://github.com/ultralytics/ultralytics/pull/25548), titled **“Replace predict feature-map dumps with class activation heatmaps.”**

Key changes include:

- Replaced raw layer-by-layer feature-map dumps with practical PyTorch class activation heatmaps using LayerCAM.
- Updated `visualize=True` during prediction to generate heatmap overlays alongside prediction outputs.
- Applied confidence and class filters, with support for visualizing a configurable number of top predictions.
- Limited activation-map generation to supported PyTorch workflows, excluding tensor sources, augmentation, and embedding workflows.
- Updated inference and documentation to reflect the new heatmap behavior.

These changes should provide more interpretable visual explanations by highlighting image regions that influence predictions, while ensuring visualizations focus on relevant detections.

The changes were implemented by [@Y-T-G](https://github.com/Y-T-G), with contributions from [@glenn-jocher](https://github.com/glenn-jocher).

To test the latest changes, install directly from the `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release.

Please test this update and share whether it resolves the issue. Thank you 🙏 for reporting it, and any further feedback is welcome if the problem persists.

**Comment by glenn-jocher** [comment]
@AnikiFan, @Y-T-G added official prediction-aware activation maps for segmentation in merged PR #25548. `model.predict(..., visualize=True)` now saves a LayerCAM overlay driven by the predicted class scores; `classes` can restrict which segmentation classes contribute.
